db/python/layers/ourdna/dashboard.py

```diff
@@ -0,0 +1,269 @@
+        """
+        _process_start_time = self.get_meta_property(sample=sample, property_name='process-start-time')
+        _process_end_time = self.get_meta_property(sample=sample, property_name='process-end-time')
+        _processing_site = self.get_meta_property(sample=sample, property_name='processing-site')
+        if (
+            _process_start_time is None
+            or _process_end_time is None
+            or _processing_site is None
+        ):
+            return None, None
+
+        processing_time = datetime.strptime(
+            _process_end_time, '%Y-%m-%d %H:%M:%S'
+        ) - datetime.strptime(
+            _process_start_time, '%Y-%m-%d %H:%M:%S'
+        )
+
+        return _processing_site, int(processing_time.total_seconds())
+
+    def get_collection_to_process_start_time(self, sample: Sample) -> int | None:
+        """
+        I want to know how long it has been since the sample was collected - SAMPLE TABLE
+        @fields: collection-time, process-start-time
+        """
+        _collection_time = self.get_meta_property(sample=sample, property_name='collection-time')
+        _process_start_time = self.get_meta_property(sample=sample, property_name='process-start-time')
+        if (
+            _collection_time is None
+            or _process_start_time is None
+        ):
+            return None
+
+        time_taken = datetime.strptime(
+            _process_start_time, '%Y-%m-%d %H:%M:%S'
+        ) - datetime.strptime(_collection_time, '%Y-%m-%d %H:%M:%S')
+
+        return int(time_taken.total_seconds())
+
+    async def query(
+        self,
+        filter_: SampleFilter,
+        check_project_ids: bool = True,
+        project_id: int = None,
+    ) -> dict:
+        """Get dashboard data"""
+        projects: set[int]
+        samples_internal: list[SampleInternal] = []
+        samples: list[Sample] = []
+        participants: list[ParticipantInternal] = []
+        sample_participants: list[ParticipantInternal] = []
+
+        # Data to be returned
+        collection_to_process_end_time: dict[str, int] = {}
+        collection_to_process_end_time_statistics: dict[str, float | None] = {}
+        collection_to_process_end_time_24h: dict[str, int] = {}
+        processing_times_by_site: dict[str, dict[int, int]] = defaultdict(
+            lambda: defaultdict(int)
+        )  # {site: {hour_bucket: count}}
+        total_samples_by_collection_event_name: dict[str, int] = defaultdict(int)
+        samples_lost_after_collection: dict[str, dict[str, Any]] = {}
+        samples_concentration_gt_1ug: dict[str, float] = {}
+        participants_consented_not_collected: list[int] = []
+        participants_signed_not_consented: list[int] = []
+
+        # TODO We should figure out if we need to handle more than a single project_id
+        # for project_id in project_ids:
+        participants.extend(
+            await self.participant_table.get_participants(project=project_id)
```

Comment:
- Why do you take a sample filter, and get all participants? Do you just want participants for the fetched samples, or actually want all. Can you add a comment either way.
- (pending the previous dot point): you can run both the participant and sample query together, in an `asyncio.gather`.
- Better to call a participant layer's `query` method, rather than interact with tables if you can avoid